File: mutagen-qa-js/src/repo/clone.js
/**
 * Shallow git-clone helper. JS port of mutagen-qa/src/mutagen/repo/clone.py.
 *
 * Contract:
 *   - Only https:// / http:// / file:// URLs are accepted. SSH and git:// are
 *     rejected (SSH needs agent forwarding in server deployments;
 *     unauthenticated git:// is deprecated).
 *   - depth = 1 by default (we generate tests from source, not history).
 *   - Timeout is enforced -- a slow remote can't wedge a job.
 *   - Destination must not exist as a non-empty directory.
 *
 * Failures throw CloneError so callers can surface a clean message instead
 * of a raw traceback.
 */

import { existsSync, mkdirSync, readdirSync, statSync } from "node:fs";
import path from "node:path";

import { runSubprocess } from "../sandbox/executor.js";

const SCHEME_ALLOW = ["https://", "http://", "file://"];
const URL_SANITY_RE = /^[A-Za-z][A-Za-z0-9+.-]*:\/\/[^\s]+$/;

export class CloneError extends Error {
  constructor(message) {
    super(message);
    this.name = "CloneError";
  }
}

function validateUrl(url) {
  if (typeof url !== "string" || !url.trim()) throw new CloneError("repo URL is empty");
  if (!URL_SANITY_RE.test(url)) throw new CloneError(`repo URL is not a valid URL: ${JSON.stringify(url)}`);
  if (!SCHEME_ALLOW.some((s) => url.startsWith(s))) {
    throw new CloneError(
      `repo URL must use https:// (or file:// for local testing); got ${JSON.stringify(url)}`,
    );
  }
}

function validateDest(dest) {
  if (existsSync(dest)) {
    const s = statSync(dest);
    if (!s.isDirectory()) throw new CloneError(`destination exists and is not a directory: ${dest}`);
    if (readdirSync(dest).length > 0) throw new CloneError(`destination is not empty: ${dest}`);
  }
}

export async function cloneRepo(url, dest, { depth = 1, timeoutS = 120 } = {}) {
  validateUrl(url);
  const resolved = path.resolve(dest);
  validateDest(resolved);
  mkdirSync(resolved, { recursive: true });

  const argv = ["git", "clone", "--depth", String(depth), "--single-branch", url, resolved];
  const res = await runSubprocess(argv, { timeoutS });
  if (res.timedOut) throw new CloneError(`git clone timed out after ${timeoutS}s`);
  if (res.returncode !== 0) {
    // Windows: EINVAL surfaces here when `git` is a package-manager .cmd shim
    // rather than a real binary. runSubprocess reports it via stderr, but the
    // message is opaque -- surface a clearer hint.
    const stderr = res.stderr || "";
    if (/spawn error.*EINVAL/i.test(stderr) || res.returncode === -4071) {
      throw new CloneError(
        `git could not be spawned (EINVAL). On Windows, ensure git.exe is on ` +
        `PATH ahead of any .cmd shim (e.g. install Git for Windows).`,
      );
    }
    const tail = (stderr || res.stdout || "").trim().slice(-800);
    throw new CloneError(`git clone failed (exit ${res.returncode}): ${tail}`);
  }
  return resolved;
}

File: mutagen-qa-py/src/mutagen/repo/clone.py
"""Shallow git clone helper.

Contract:
    - HTTPS git URLs and ``file://`` URLs are accepted. SSH / git:// are
      rejected -- SSH would need agent forwarding in server deployments,
      and unauthenticated git:// is deprecated.
    - Depth is 1 by default. Test-generation doesn't need history; the
      clone is a snapshot.
    - Timeout is enforced. A slow / hanging remote can't wedge a job.
    - Destination must not already exist as a non-empty directory.

Failures raise ``CloneError`` so the caller can surface a clean message to
the SSE stream instead of a raw traceback.
"""

from __future__ import annotations

import re
import shutil
import subprocess
from pathlib import Path

# Loose but useful: catches obvious garbage (``javascript:``, ``sh://``,
# spaces, no scheme at all) without banning legitimate mirrors. The
# scheme allowlist is the safety net; regex is the smoke test.
_SCHEME_ALLOW = ("https://", "http://", "file://")
_URL_SANITY_RE = re.compile(r"^[A-Za-z][A-Za-z0-9+.-]*://[^\s]+$")


class CloneError(RuntimeError):
    """Any failure surfaced to the caller: bad URL, git not on PATH, timeout,
    non-empty destination, or a non-zero git exit code."""


def _validate_url(url: str) -> None:
    if not isinstance(url, str) or not url.strip():
        raise CloneError("repo URL is empty")
    if not _URL_SANITY_RE.match(url):
        raise CloneError(f"repo URL is not a valid URL: {url!r}")
    if not url.startswith(_SCHEME_ALLOW):
        raise CloneError(
            f"repo URL must use https:// (or file:// for local testing); got {url!r}"
        )


def _validate_dest(dest: Path) -> None:
    if dest.exists():
        if not dest.is_dir():
            raise CloneError(f"destination exists and is not a directory: {dest}")
        # ``.git`` presence would mean we're overlapping an existing clone --
        # git would refuse anyway, but this gives a clearer message.
        if any(dest.iterdir()):
            raise CloneError(f"destination is not empty: {dest}")


def clone_repo(
    url: str,
    dest: Path,
    *,
    depth: int = 1,
    timeout_s: int = 120,
) -> Path:
    """Shallow-clone ``url`` into ``dest``. Returns the destination path.

    ``depth=1`` is a deliberate default: we generate tests from source, not
    from history. If you ever need history for a specific analysis, override
    at the call site rather than here.
    """
    _validate_url(url)
    dest = dest.resolve()
    _validate_dest(dest)
    dest.mkdir(parents=True, exist_ok=True)

    git = shutil.which("git")
    if not git:
        raise CloneError("git CLI not found on PATH")

    cmd = [git, "clone", "--depth", str(depth), "--single-branch", url, str(dest)]
    try:
        proc = subprocess.run(
            cmd, capture_output=True, text=True, encoding="utf-8",
            errors="replace", timeout=timeout_s, check=False,
        )
    except subprocess.TimeoutExpired as e:
        # Leave a partial clone visible for debugging but don't retry --
        # a slow remote at N seconds is likely slow at 2N seconds.
        raise CloneError(f"git clone timed out after {timeout_s}s") from e

    if proc.returncode != 0:
        # git writes progress to stderr; truncate so the error message
        # stays readable in the SSE stream / UI.
        tail = (proc.stderr or proc.stdout or "").strip()[-800:]
        raise CloneError(f"git clone failed (exit {proc.returncode}): {tail}")
    return dest
